### DISCUSSION TITLE
Increasing rule size & rule sets

### DISCUSSION THREAD

**Discussion by rapi3** [arendst/Tasmota#11761] [Q&A] [answered]
Hello,

I start compiling my custom firmware with only the features & sensors I use and disable everything else I can, at this moment I come to this configuration:

![Screenshot_2021-04-17_16-31-56](https://user-images.githubusercontent.com/18531150/115116584-8af4a180-9f9a-11eb-9455-ac711d5be334.png) ![Screenshot_2021-04-17_16-32-14](https://user-images.githubusercontent.com/18531150/115116587-8f20bf00-9f9a-11eb-83fa-ecf32028bb9b.png)

For example after few days up I have available ram on:
- Wemos D1 mini with different configurations; DS18x20, Buzzer, IR, relay, switch... : 29.3Kb, 28.1Kb, 26.3Kb, 21.7Kb
- RF Bridge: 27.1Kb ( with DS18x20 temp sensor ) and 23,9Kb ( with IR bridge & DS18x20 temp sensor )

Now I try to use rules in Tasmota at maximum to avoid dependency from other system ( HA, MQTT... ) in case of components failure but I come quick to max rule limit even if I am using  compression:
`#define USE_UNISHOX_COMPRESSION // Add support for string compression in Rules or Scripts`

I can see in tasmota.h this lines:
```
const uint8_t MAX_RULE_SETS = 3;            // Max number of rule sets of size 512 characters
const uint16_t MAX_RULE_SIZE = 512;         // Max number of characters in rules
```

**_How safe it is to increase them; and to what value?
What is the best practice for this ?_**

for example:
`MAX_RULE_SIZE = 1024;         // Max number of characters in rules`
or
`MAX_RULE_SETS = 6;            // Max number of rule sets of size`

thank you.

**Comment by arendst** [comment, marked as answer]
> How safe it is to increase them;

Very unsafe as it breaks the fixed settings flash area of exactly 4k

>  and to what value?

See above

> What is the best practice for this ?

Learn to use simpler rules or switch to script language.

**Reply by sfromis** [reply]
I can recommend stepping up to ESP32, where you have the Berry scripting language, which also works out better for stuff complex enough for you to not fit into the over 3k effective (before compression) space available for rules. Or do such automations backend, I'm finding Node-Red very nice to use.

**Comment by rapi3** [comment]
I have to look at scripting some day... when I feel brave enough.

In general my rules are simple enough but repeating specially on RF bridge where I plan to use more RF buttons to toggle lights and do automation, imagine having 30 rules like this only for RF to turn on/off the lights:
`ON RfReceived#Data=968651 DO Publish Wemos/cmnd/W10/power2 2 ENDON`
+ other stuff for IR...

So it is more a flash storage problem and not RAM limitation ?
What about using Universal File System to store them it will help if we can store rules in files there something like rule1.bat rule2.bat rule3.bat... ?

**Reply by barbudor** [reply]
Sonoff RFbridge is using an ESP8285 with only 1MB flash, not supporting Universal file system
So better look into NodeRed or other external tool

**Reply by rapi3** [reply]
True but at least for wemos D1 mini where we have 2Mb UFS available.

**Reply by sfromis** [reply]
It is not even a "flash storage problem", but an issue about how much can be stored within the fixed 4K block of settings, housing lots of different items.

You can to some extent shorten your rules by storing some strings in variables (var or mem), `var1 Publish Wemos/cmnd`, and then use them in the DO/ENDON part to do things like 
`ON RfReceived#Data=968651 DO %var1%/W10/power2 2 ENDON`
As this defeats compression, you'll probably gain less than the fewer characters input.

In the file system, you can store commands to be executed from a rule (using `UfsRun`), but not rules where the triggers will have to be evaluated all the time.

**Reply by rapi3** [reply]
Thank you for the UfsRun tip.

**Reply by barbudor** [reply]
And outside of flash, RAM also is a problem. Even if you where considering to implement Rule4..+ in RAM only (and get them initialized from a file) that could be too much.
So replace your WemosD1 by ESP32 and start using Berry

**Reply by sfromis** [reply]
Another potential rule-shortening tip for such repetitive rules...
`ON RfReceived#Data do event r=%value% ENDON`
`ON Event#r=968651  DO ...`
Again the compression means that you'll likely not get "full credit".

**Reply by rapi3** [reply]
ok thanks for tips, now I also learn that I have to do some design changes to mqtt full topic to save on characters and remove the `Wemos/` part.

**Reply by sfromis** [reply]
Not tested this idea, but you may be able to get around the rule size issue when having the file system.
`ON RfReceived#Data do UfsRun rf%value%.bat ENDON`
This would take a lot of small files, and I'm not sure how much minimal space per file is.

**Reply by barbudor** [reply]
You can also have a single rule such as `ON RfReceived#Data do publish rfreceived/%value% ENDON`
And have any concerned device to subscribe to the `rfReceived/XXXXX` it needs

**Reply by sfromis** [reply]
Of course, concerned devices could also "just" subscribe to the existing RfReceived message, without repackaging it at all.

**Reply by rapi3** [reply]
I think this will add a lot of mqtt traffic with RF noise and moving filtering to devices that turn on/off the light or doing other stuff.

**Reply by rapi3** [reply]
I already have made auto-discovery IR+RF sensor defined in this devices to move automation on HA. I can move this to a vemos device in ufs command files to save some space from rules on RF:
`ON mqtt#connected DO Publish2 homeassistant/sensor/%topic%/IR/config {"name":"wemos-05 IR-REC","stat_t":"Wemos/tele/W05/RESULT","avty_t":"Wemos/tele/W05/LWT","pl_avail":"Online","pl_not_avail":"Offline","uniq_id":"7CAA7C_IR_REC","dev":{"ids":["7CAA7C"]},"val_tpl":"{{value_json.IrReceived.Data}}"} ENDON
`
`ON mqtt#connected DO Publish2 homeassistant/sensor/RF-01/RF/config {"name":"Bridge RF 01 433-REC","stat_t":"Bridge/tele/RF-01/RESULT","avty_t":"Bridge/tele/RF-01/LWT","pl_avail":"Online","pl_not_avail":"Offline","uniq_id":"EF5544_433_REC","dev":{"ids":["EF5544"]},"val_tpl":"{{value_json['RfReceived']['Data']}}"} ENDON`

**Reply by sfromis** [reply]
I'd not expect %-variables to be expanded from a UFS .bat file. They only work within rules.

**Reply by rapi3** [reply]
yes no problem to name it full as is enough space in ufs - 2Mb.
